iree/compiler/Conversion/Common/LinalgBufferizePass.cpp

```diff
@@ -49,458 +50,528 @@ namespace mlir {
-    // Leave the previous transfer_write to dead code as it still has uses at
-    // this point.
-    b.create<vector::TransferWriteOp>(
-        loc, writeOp.vector(), newInputBuffer, writeOp.indices(),
-        writeOp.permutation_map(), writeOp.mask(),
-        writeOp.in_bounds() ? *writeOp.in_bounds() : ArrayAttr());
+    return nullptr;
   }
-  return success();
+  return allocationFn(b, loc, resultType.getShape(),
+                      resultType.getElementType(), dynamicDims);
 }
 
-// Extract int64_t values from the assumed ArrayAttr of IntegerAttr.
-static SmallVector<int64_t, 4> extractFromI64ArrayAttr(Attribute attr) {
-  return llvm::to_vector<4>(llvm::map_range(
-      attr.cast<ArrayAttr>(),
-      [](Attribute a) -> int64_t { return a.cast<IntegerAttr>().getInt(); }));
+template <typename TensorType>
+static MemRefType getMemrefTypeForTensor(TensorType tensorType,
+                                         ArrayRef<AffineMap> layout = {},
+                                         unsigned memorySpace = 0) {
+  return MemRefType::get(tensorType.getShape(), tensorType.getElementType(),
+                         layout, memorySpace);
 }
 
-LogicalResult convertInterfaceLoadTensorOp(
-    OpBuilder &b, IREE::Flow::DispatchTensorLoadOp loadOp,
-    BlockAndValueMapping &bvm) {
-  OpBuilder::InsertionGuard g(b);
-  b.setInsertionPoint(loadOp);
-  Location loc = loadOp.getLoc();
-  Value memref = bvm.lookup(loadOp.source());
-  Value res = createSubviewOp(b, loc, memref, loadOp.getMixedOffsets(),
-                              loadOp.getMixedSizes(), loadOp.getMixedStrides());
-
-  bvm.map(loadOp.result(), res);
-  transferShapeOpsToMemref(b, loadOp.result(), res, bvm);
-  return success();
+/// Creates a subview operation given the `src`, `offsets`, `sizes` and
+/// `strides`. Handles the corner case where the `offsets`, `sizes` and
+/// `strides` are empty in which case just forward the `src` value.
+/// TODO(ataei): Instead create memref.subview %v [][][] folder.
+static Value createSubviewOp(OpBuilder &b, Location loc, Value src,
+                             ArrayRef<OpFoldResult> offsets,
+                             ArrayRef<OpFoldResult> sizes,
+                             ArrayRef<OpFoldResult> strides,
+                             MemRefType resultType = MemRefType()) {
+  if (offsets.empty() && sizes.empty() && strides.empty()) return src;
+  return b.create<memref::SubViewOp>(loc, resultType, src, offsets, sizes,
+                                     strides);
 }
 
+//===----------------------------------------------------------------------===//
+// There might be cases when the `value` stored into a
+// `flow.dispatch.tensor.store` operation is obtained from operation that
+// computes the value (say a `linalg` operation) through a series of `reshapes`,
+// `cast` etc. When trying to reuse the buffer for the result passed in to the
+// dispatch region for these operations, these operations need to be "replayed"
```

Comment:
Or instead of deciding on the type we can check total size is the same (e.g can acquire the whole buffer) we can reuse by inserting a reshape.

Reply:
That only works for static shapes. Plus just reshapes might not be enough. Some casts might be needed too. Would be hard to figure it out just based on the final type. Easiest is to replay the instructions in reverse.